### PR TITLE
Fix expression conversion

### PR DESCRIPTION
**Pull request by UnderscoreTud**
### Description
This PR fixes an issue where the parser wouldn't attempt to convert an expression if `Expression#canReturn(Class)` returns `true`. Causing issues when an expression *could* return a certain type but doesn't once evaluated. For instance, adding two unset variables, which can potentially return variety of types (such as a string), and using the expression in a pattern that expects strings (e.g., `formatted ({_foo} + {_bar})` would result in a [stack trace](https://pastebin.com/C9h1LrCh) being thrown because the array returned from the arithmetic expression wasn't converted to a string array.

---
**Target Minecraft Versions:** any 
**Requirements:** none 
**Related Issues:** #7164

### CODE REVIEW

File: src/main/java/ch/njol/skript/lang/SkriptParser.java [hunk 1]
```diff
@@ -371,8 +371,13 @@ private <T> Expression<? extends T> parseSingleExpr(boolean allowUnparsedLiteral
...
 						if (parsedExpression.canReturn(type)) {
-							log.printLog();
-							return (Expression<? extends T>) parsedExpression;
+							Expression<? extends T> convertedExpression = parsedExpression.getConvertedExpression(type);
+							if (convertedExpression != null) {
+								log.printLog();
+								return convertedExpression;
+							}
+							log.printError(parsedExpression.toString(null, false) + " " + Language.get("is") + " " + notOfType(type), ErrorQuality.NOT_AN_EXPRESSION);
+							return null;
```
Comment (sovdeeth) [inline src/main/java/ch/njol/skript/lang/SkriptParser.java:380]: do you think it's worth making a helper method since this is duplicated?

**Review by sovdeeth — APPROVED** [review]

**Review by Moderocky — APPROVED** [review]

**Review by APickledWalrus — CHANGES_REQUESTED** [review]
Will review in depth soon. tud and I discussed some potential issues/edge cases that still need fixed

File: src/main/java/ch/njol/skript/lang/SkriptParser.java [hunk 2]
```diff
@@ -540,13 +539,14 @@ private Expression<?> parseSingleExpr(boolean allowUnparsedLiteral, @Nullable Lo
...
+					Class<?> parsedReturnType = parsedExpression.getReturnType();
 					for (int i = 0; i < types.length; i++) {
 						Class<?> type = types[i];
 						if (type == null) // Ignore invalid (null) types
 							continue;
 
-						// Check return type against everything that expression accepts
-						if (parsedExpression.canReturn(type)) {
+						// Check return type against the expression's return type
+						if (type.isAssignableFrom(parsedReturnType)) {
```
Comment (sovdeeth) [inline src/main/java/ch/njol/skript/lang/SkriptParser.java:549]: what's the logic for not using canReturn?

Comment (sovdeeth) [reply src/main/java/ch/njol/skript/lang/SkriptParser.java:549]: oh i think that was explained in the original pr

**Review by sovdeeth — APPROVED** [review]

**Comment by APickledWalrus** [comment]
I've gone ahead and moved the logic out of SimpleExpression and into ConvertedExpression. This is to preserve the behavior of expressions overriding the method AND that the logic/handling for this is consistent across Skript. The `canReturn` logic has been moved out of SkriptParser, as its usage there is an incorrect assertion (an object that might be `Expression<A>` or `Expression<B>`  is NOT the same as `Expression<A>`, so we cannot return it as one). It has to be handled by ConvertedExpression so that we can ensure arrays are typed correctly (the root problem).

**Review by Moderocky — APPROVED** [review]

**Pull request merged**